[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112486
* #112551
* __->__ #112550

**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -48,12 +48,24 @@
     KeywordArg("w_dtype"),
 )
 
+dequantize_per_channel_to_bf16_weight_pattern = CallFunction(
+    prims.convert_element_type.default,
+    dequantize_per_channel_weight_pattern,
+    KeywordArg("autocast_weight_convert_dtype"),
+)
```
maybe use a util function like `dtype_convert(pattern, dtype)` to simplify the places that need a type conversion on the output of the pattern?

leslie-fang-intel replied:
Thanks for the suggestion. Added `_generate_pattern_with_dtype_convert` as the util function to append pattern with dtype conversion.

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering"


**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering"


**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering"


**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering"


**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @eellison @jerryzh168, Could you kindly help to review these 3 PRs in this ghstack for the Inductor lowering of int8-mixed-bf16 quantization case?

eellison commented:
@jerryzh168 want to review this one ?

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -1162,16 +1264,26 @@ def _generate_qlinear_weight_prepack_patterns():
-        # Register to pass_number 1, so we can do dequant promotion in pass_number 0.
-        _register_qconv_weight_prepack_pass(weight_prepack_pattern, pass_number=1)
+    for dtype in [torch.float32, torch.bfloat16]:
+        # Step 1: Dequant promotion for int8-mixed-fp32/bf16
```
what does dequant promotion mean? can you write down the patterns or transformation?

leslie-fang-intel replied:
Thanks for the suggestion. Add the patterns and description for this pass.

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -1162,16 +1264,26 @@ def _generate_qlinear_weight_prepack_patterns():
+    for dtype in [torch.float32, torch.bfloat16]:
+        # Step 1: Dequant promotion for int8-mixed-fp32/bf16
+        _register_dequant_promotion_pass(
+            dequantize_per_tensor_activation_pattern
+            if dtype == torch.float32
+            else _generate_pattern_with_dtype_convert(
+                dequantize_per_tensor_activation_pattern,
+                KeywordArg("autocast_act_dtype"),
```
can you just pass in dtype as an argument for `_generate_pattern_with_dtype_convert` and remove the if check?

leslie-fang-intel replied:
Thanks for the comments. The docs added and also the link to RFC for detail design. I will also update the tutorial for the usage of this feature.

leslie-fang-intel replied:
Thanks and changed. It saves lots of redundant code.

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -146,7 +167,12 @@ def generate_pattern_with_output_quant(computation_call):
                         CallFunction(
                             aten.mul.Tensor,
-                            computation_call,
+                            computation_call
+                            if dtype == torch.float32
+                            else _generate_pattern_with_dtype_convert(
+                                computation_call,
+                                KeywordArg("autocast_output_quant_dtype"),
+                            ),
```
looks like this appears multiple times, refactoring a bit would be better I think

leslie-fang-intel replied:
Same as above and changed accordingly.

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -18,6 +18,15 @@
 quantized_decomposed = torch.ops.quantized_decomposed
 quantized = torch.ops.quantized
 
+
```
could you add some docs about how bfloat16 is handled (and its interaction with quantization)?

leslie-fang-intel replied:
Thanks for the comments. The docs added and also the link to RFC for detail design. I will also update the tutorial for the usage of this feature.

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -756,30 +787,42 @@ def _register_quantization_lowerings():
-        # dequant pattern has more than 1 users to be promoted
-        return True
-    return False
+def _is_valid_dequant_promotion_pattern(dtype=torch.float32):
```
could you add some docs to explain dequant_promotion

leslie-fang-intel replied:
Thanks for the suggestions. Docs added as in https://github.com/pytorch/pytorch/pull/112550#discussion_r1388459772

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -448,29 +448,38 @@ def __init__(
-                counters["inductor"]["qconv2d_weight_prepack_matcher_nodes"], 6
-            )
+        for int8_mixed_bf16 in (
+            [False, True] if torch.ops.mkldnn._is_mkldnn_bf16_supported() else [False]
```
will it be better to split this to two tests?

leslie-fang-intel replied:
Thanks for the suggestion and split.

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -488,33 +497,35 @@ def __init__(
-            #    [qconv2d_pointwise_default, relu]
-            self.assertEqual(counters["inductor"]["qconv2d_unary_matcher_count"], 1)
-            self.assertEqual(counters["inductor"]["qconv2d_unary_matcher_nodes"], 2)
+        for int8_mixed_bf16 in (
```
same here

leslie-fang-intel replied:
Thanks for the suggestion and split.

jerryzh168 reviewed (approved):
LG overall, please add more some docs

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering"


**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
